### PR TITLE
Add amx_searchcmd command in adminhelp plugin and do some clean up

### PR DESCRIPTION
**Pull request by Arkshine**
Essentially:
- Added `amx_searchcmd`: Similar to `amx_help` except you can provide a string to match partially the command names. Quite useful if you have a lot of commands.
- Moved the hardcoded `15` seconds for the help message to a new`amx_help_display_msg_time` cvar.
- The commands description are translated.
- Some general cleanup, using new stuff as well.

Feel free to correct the english if necessary, and to provide the translations in your language.

### CODE REVIEW

File: plugins/adminhelp.sma [hunk 1]
```diff
@@ -13,133 +13,181 @@
 
 #include <amxmodx>
 
-new g_cvarDisplayClientMessage;
-new g_cvarHelpAmount;
+const MaxMapLength         = 32;
```
Comment (IgnacioFDM) [inline plugins/adminhelp.sma]: Should be 64

Comment (rsKliPPy) [inline plugins/adminhelp.sma]: If you take a look around ReHLDS, it looks like it is indeed 32. If you find proof that it's 64, link it.

Comment (IgnacioFDM) [inline plugins/adminhelp.sma]: Well I was mistaken because changelevel uses 64 chars. However if you look at  

https://github.com/dreamstalker/rehlds/blob/94fb06f95b4e1e9f150cd4e8c0f4e0fb86e552e2/rehlds/engine/pr_cmds.cpp#L1549

you'll see it's 32 as you said, therefore I believe ```MaxMapLength``` should be 33 (or array be ```MaxMapLength + 1```). Just tested in a server and 32 char maps do work.

Comment (rsKliPPy) [inline plugins/adminhelp.sma]: But if you search through ReHLDS for mapname, in some places it's defined with 32 (a null-terminated string). I'm confused.

Comment (IgnacioFDM) [inline plugins/adminhelp.sma]: It's odd indeed, and oddly works well as far as I can tell, it downloaded the map correctly and I can play correctly. If I remember from memory, the important parts are [40] or [64] since they include ```maps/``` and ```.bsp```

**Comment by Freeman-AM** [comment]
I have some questions for you Arkshine, nothing big, don't worry :
Why didn't you translate SEARCH_CMD_INFO and NO_MATCHING_RESULTS in French?

Did you create or plan to create a topic on AM asking for translations ?

I started a project, adding UTF8 support for amxmodx french dictionnary. How github deals with conflicting PR ?

**Comment by Arkshine** [comment]
Missing keys can be dealt later. It's expected at this point to not be able to fill for every language. Just forgot for french. If someone wants to add his language, feel free to do it now.

If a PR is detected as not-mergeable by Github, you will have to fix manually the conflict.

**Comment by Arkshine** [comment]
@rsKliPPy @IgnacioFDM  @Ni3znajomy @Nextra Can someone do some fast review please?

**Comment by Freeman-AM** [comment]
From my point of view, the code looks good. But i prefer you have the point of people who have more knowledge than me.

I have some question, why don't you use create_cvar instead of register_cvar ?

When you are at optimization, why don't you use that sort of bit code for the boolean ?
new HasAwp
#define SetBitHasAwp(%1)		HasAwp |= 1<<(%1&31)
#define RemoveBitHasAwp(%1)		HasAwp &= ~(1<<(%1&31))
#define HasBitHasAwp(%1)		HasAwp & 1<<(%1&31)

**Comment by IgnacioFDM** [comment]
Not really directed at you, @Freeman-AM, but more as general statement since many people think bitfields are a good idea, when actually **they are not**

- Bitfields are more annoying/more "complex" than plain arrays
- Memory savings are simply irrelevant, unless you have hundreds of thousands of elements. We are talking about bytes, when computers are in the order of gigabytes. HLDS by itself runs at about 150MB. Let's say you convert 200 arrays of 33 elements into 200 bitfields. You just saved 25KB of RAM. Do you think 25KB of RAM matters in a 150MB program?
- Bitfields are not faster. In fact they may even be slower.

See the following "benchmark"

```C
new g_HasAwp[33];
new HasAwp;
#define SetBitHasAwp(%1)	(HasAwp |= 1<<(%1&31))
#define RemoveBitHasAwp(%1)	(HasAwp &= ~(1<<(%1&31)))
#define HasBitHasAwp(%1)	(HasAwp & 1<<(%1&31))
new Float:start, Float:end;

start = get_sysgametime(); // Returns game time based on the systemclock this instant
for(new i = 0; i < 10000000; i++)
{
	// Unroll to reduce for loop overhead
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
	SetBitHasAwp(2);
}
end = get_sysgametime();
server_print("Set bitfield: %f seconds", end - start);

start = get_sysgametime();
for(new i = 0; i < 10000000; i++)
{
	// Unroll to reduce for loop overhead
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
	g_HasAwp[2] = true;
}
end = get_sysgametime();
server_print("Set array: %f seconds", end - start);

new dummy;
start = get_sysgametime();
for(new i = 0; i < 10000000; i++)
{
	if(HasBitHasAwp(2))	dummy++;
	if(HasBitHasAwp(2))	dummy++;
	if(HasBitHasAwp(2))	dummy++;
	if(HasBitHasAwp(2))	dummy++;
	if(HasBitHasAwp(2))	dummy++;
	if(HasBitHasAwp(2))	dummy++;
	if(HasBitHasAwp(2))	dummy++;
	if(HasBitHasAwp(2))	dummy++;
	if(HasBitHasAwp(2))	dummy++;
	if(HasBitHasAwp(2))	dummy++;
	if(HasBitHasAwp(2))	dummy++;
	if(HasBitHasAwp(2))	dummy++;
	if(HasBitHasAwp(2))	dummy++;
	if(HasBitHasAwp(2))	dummy++;
	if(HasBitHasAwp(2))	dummy++;
	if(HasBitHasAwp(2))	dummy++;
}
end = get_sysgametime();
server_print("Check bitfield %f seconds (dummy %d)", end - start, dummy);

dummy = 0;
start = get_sysgametime();
for(new i = 0; i < 10000000; i++)
{
	if(g_HasAwp[2]) dummy++;
	if(g_HasAwp[2]) dummy++;
	if(g_HasAwp[2]) dummy++;
	if(g_HasAwp[2]) dummy++;
	if(g_HasAwp[2]) dummy++;
	if(g_HasAwp[2]) dummy++;
	if(g_HasAwp[2]) dummy++;
	if(g_HasAwp[2]) dummy++;
	if(g_HasAwp[2]) dummy++;
	if(g_HasAwp[2]) dummy++;
	if(g_HasAwp[2]) dummy++;
	if(g_HasAwp[2]) dummy++;
	if(g_HasAwp[2]) dummy++;
	if(g_HasAwp[2]) dummy++;
	if(g_HasAwp[2]) dummy++;
	if(g_HasAwp[2]) dummy++;
}
end = get_sysgametime();
server_print("Check array %f seconds (dummy %d)", end - start, dummy);

dummy = 0;
start = get_sysgametime();
for(new i = 0; i < 10000000; i++)
{
	if(!HasBitHasAwp(2)) dummy++;
	if(!HasBitHasAwp(2)) dummy++;
	if(!HasBitHasAwp(2)) dummy++;
	if(!HasBitHasAwp(2)) dummy++;
	if(!HasBitHasAwp(2)) dummy++;
	if(!HasBitHasAwp(2)) dummy++;
	if(!HasBitHasAwp(2)) dummy++;
	if(!HasBitHasAwp(2)) dummy++;
	if(!HasBitHasAwp(2)) dummy++;
	if(!HasBitHasAwp(2)) dummy++;
	if(!HasBitHasAwp(2)) dummy++;
	if(!HasBitHasAwp(2)) dummy++;
	if(!HasBitHasAwp(2)) dummy++;
	if(!HasBitHasAwp(2)) dummy++;
	if(!HasBitHasAwp(2)) dummy++;
	if(!HasBitHasAwp(2)) dummy++;
}
end = get_sysgametime();
server_print("Check ! bitfield %f seconds (dummy %d)", end - start, dummy);

dummy = 0;
start = get_sysgametime();
for(new i = 0; i < 10000000; i++)
{
	if(!g_HasAwp[2]) dummy++;
	if(!g_HasAwp[2]) dummy++;
	if(!g_HasAwp[2]) dummy++;
	if(!g_HasAwp[2]) dummy++;
	if(!g_HasAwp[2]) dummy++;
	if(!g_HasAwp[2]) dummy++;
	if(!g_HasAwp[2]) dummy++;
	if(!g_HasAwp[2]) dummy++;
	if(!g_HasAwp[2]) dummy++;
	if(!g_HasAwp[2]) dummy++;
	if(!g_HasAwp[2]) dummy++;
	if(!g_HasAwp[2]) dummy++;
	if(!g_HasAwp[2]) dummy++;
	if(!g_HasAwp[2]) dummy++;
	if(!g_HasAwp[2]) dummy++;
	if(!g_HasAwp[2]) dummy++;
}
end = get_sysgametime();
server_print("Check ! array %f seconds (dummy %d)", end - start, dummy);

```

These are the results on my machine (tm)
```
Set bitfield: 5.941339 seconds
Set array: 5.701327 seconds
Check bitfield 7.905452 seconds (dummy 160000000)
Check array 7.856451 seconds (dummy 160000000)
Check ! bitfield 3.056175 seconds (dummy 0)
Check ! array 3.034175 seconds (dummy 0)
```

In other words, they are the same. Maybe even slower than arrays. It doesn't really matter, basically **anything** else will dominate in terms of CPU time compared to setting or checking an array cell or bitfield.

**TL;DR: Use arrays, not bitfields.**

**Comment by Freeman-AM** [comment]
Thanks for breaking some preconcepted ideas i read all along by many people on AM. It wasn't for memory saving i switched over bitfield but mostly because for many people, the statement is it's better than an array concerning CPU usage. So i automated their usages everywhere.

Let's back to the topic, from my previous post :

> From my point of view, the code looks good. But i prefer you have the point of people who have more knowledge than me. Why don't you use create_cvar instead of register_cvar ?

**Comment by rsKliPPy** [comment]
Me gusta. :sheep:
